src/ops/nn.jl

```diff
@@ -137,51 +137,65 @@ function rnn(cell, inputs; initial_state=nothing, dtype=nothing, sequence_length
 * `parallel_iterations`: Number of iterations to run in parallel, defaulting to `32`. Ops that have no temporal dependency can be run in parallel and will be. Trades time for space - larger values use more memory.
 * `swap_memory`: Optional `Bool`, which if `true` allows transparent swapping of `Tensor`s needed for back propagation from the CPU to the GPU and vice versa. Allows training of RNNs which wouldn't fit on a single GPU. Defaults to `false`.
 * `time_major`: Shape format for `inputs` and `outputs` `Tensor`s. Determines whether the first dimension of each is `max_time` (`true`) or `batch_size` (`false`, default). `true` is more efficient but is the transpose of most TensorFlow operations.
 * `scope`: `VariableScope` for the subgraph. Defaults to `RNN`.
 """
-@op function dynamic_rnn(cell, inputs; sequence_length=nothing, initial_state=nothing, dtype=nothing, parallel_iterations=nothing, swap_memory=false, time_major=false, scope="RNN")
-    sequence_length === nothing || error("sequence_length parameter not supported yet")
-    time_major == false || error("Time-major order not supported yet")
-    time_step = tf.constant(1)
-    num_steps = convert(Tensor{Int64}, tf.shape(inputs)[2])
+@op function dynamic_rnn(cell, inputs, sequence_length=nothing; initial_state=nothing, dtype=nothing, parallel_iterations=nothing, swap_memory=false, time_major=false, scope="RNN")
+    #TODO Make this all work with non-3D inputs
+
+    if time_major
+        inputs=permutedims([2,1,3])
+    end
 
     if initial_state === nothing
         if dtype === nothing
             error("dtype must be set if initial_state is not provided")
         end
-        batch_size = get_shape(first(inputs), 1)
+        batch_size = get_shape(inputs, 1)
         initial_state = zero_state(cell, batch_size, dtype)
     end
 
     state = initial_state
     input_dim = get_shape(inputs, 3)
-
     output = tf.zeros(Tensor, eltype(state), (get_shape(inputs, 1), output_size(cell)))
 
+    time_step = tf.constant(1)
+    num_steps = convert(Tensor{Int64}, tf.shape(inputs)[2])
+
     while_output = @tf while time_step ≤ num_steps
         slice_start = tf.stack([0, time_step-1, 0])
         slice_size = tf.stack([-1, 1, -1])
         data = tf.slice(inputs, slice_start, slice_size)
         data = tf.squeeze(data, [2])
+        #TODO: Once indexing supports it, the whole of the above can be replaced with: `data=inputs[:, timestep, :]`
         local new_state
+        new_output = output
+
+
         variable_scope(scope) do
-            output, new_state = cell(data, state, input_dim)
+            new_output, new_state = cell(data, state, input_dim)
+            if !isa(sequence_length, Void) #This should be removed by the julia lowering process
+                #Only update output and state for rows that are not yet passed their ends
+                have_passed_end = sequence_length .< time_step
```

Comment:
`time_step` is a scalar.

`sequence_length` is normally a vector -- you get to specify 1 per element in the batch.

It can be a scalar, but this is not the most effective code for that case -- instead in that case should take a slice before starting the `dynamic_rnn`.